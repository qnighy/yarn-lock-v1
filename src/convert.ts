import yaml from "js-yaml";

export function convertLockfile(lockV2: string): string {
  const lockV2Obj = yaml.load(lockV2);

  let lockV1 = "";
  lockV1 += "# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.\n"
  lockV1 += "# yarn lockfile v1\n"
  lockV1 += "\n";

  for (const [packageKey, packageData] of Object.entries(lockV2Obj as object)) {
    if (packageKey === "__metadata") continue;
    if (/@(?:patch|workspace):/.test(packageKey)) continue;

    lockV1 += "\n";
    lockV1 += `${convertKey(packageKey)}:\n`;
    if (typeof packageData.version === "string") {
      lockV1 += `  version ${JSON.stringify(packageData.version)}\n`;
    }
    if (typeof packageData.resolution === "string") {
      const resolvedUrl = convertResolution(packageData.resolution);
      if (resolvedUrl) {
        lockV1 += `  resolved ${JSON.stringify(resolvedUrl)}\n`;
      }
    }
    lockV1 += convertDependencies(packageData.dependencies);
  }

  return lockV1;
}

function convertDependencies(dependencies: unknown): string {
  let output = "";
  if (typeof dependencies !== "object" || dependencies === null) return output;
  output += '  dependencies:\n';

  for (const [depKey, depValue] of Object.entries(dependencies)) {
    output += `    ${escapeIfNeeded(depKey)} ${JSON.stringify(depValue)}\n`;
  }
  return output;
}

function convertResolution(resolutionV2: string): string | null {
  const [packageName, packageResolution] = resolutionV2.split(/(?!^@)@/, 2);
  const packageBaseName = packageName.split('/').reverse()[0];
  const [protocol, protocolValue] = packageResolution.split(':', 2);
  switch (protocol) {
    case 'npm': {
      const version = protocolValue;
      return `https://registry.yarnpkg.com/${packageName}/-/${packageBaseName}-${version}.tgz`;
    }
    default:
      throw new Error(`unknown package protocol: ${JSON.stringify(protocol)}`);
  }
}

function convertKey(keyV2: string): string {
  return keyV2.split(", ").map((constraintV2) => {
    const constraintV1 = constraintV2.replace("@npm:", "@");
    return escapeIfNeeded(constraintV1);
  }).join(", ");
}

function escapeIfNeeded(val: string): string {
  return /^@| /.test(val) ? JSON.stringify(val) : val;
}
