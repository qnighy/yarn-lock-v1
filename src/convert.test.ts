import { describe, expect, it, jest } from "@jest/globals";
import fs from "fs";
import path from "path";
import { setupRecorder } from "nock-record";
import nodeFetch, { Headers, Response } from "node-fetch";

import { convertLockfile, Fetcher } from "./convert";

const record = setupRecorder({
  mode: process.env.NOCK_BACK_MODE === "record" ? "record" : "lockdown",
});

describe("convertLockfile", () => {
  it("convers normal dependencies", async () => {
    const { completeRecording, assertScopesFinished } = await record(
      "yarn-lock1"
    );

    const lockV2 = fs.readFileSync(
      path.resolve(__dirname, "./__fixtures__/yarn-lock1.txt"),
      "utf-8"
    );
    const lockV1 = await convertLockfile(lockV2);
    expect(lockV1).toMatchSnapshot();

    completeRecording();
    assertScopesFinished();
  });

  it("converts confluent dependencies", async () => {
    const { completeRecording, assertScopesFinished } = await record(
      "yarn-lock2"
    );

    const lockV2 = fs.readFileSync(
      path.resolve(__dirname, "./__fixtures__/yarn-lock2.txt"),
      "utf-8"
    );
    const lockV1 = await convertLockfile(lockV2);
    expect(lockV1).toMatchSnapshot();

    completeRecording();
    assertScopesFinished();
  });

  it("falls back to offline inference if fetch failed", async () => {
    const fetch: Fetcher = async (url) => {
      if (/is-regex\/1\.1\.2/.test(url)) {
        return new Response("", { status: 503, url });
      } else {
        return nodeFetch(url);
      }
    };
    const interaction = {
      warn: jest.fn<string, []>(),
    };
    const { completeRecording, assertScopesFinished } = await record(
      "yarn-lock1-offline"
    );

    const lockV2 = fs.readFileSync(
      path.resolve(__dirname, "./__fixtures__/yarn-lock1.txt"),
      "utf-8"
    );
    const lockV1 = await convertLockfile(lockV2, { fetch, interaction });
    expect(lockV1).toMatchSnapshot();

    completeRecording();
    assertScopesFinished();

    expect(interaction.warn.mock.calls).toEqual([
      [
        "Error reading is-regex/1.1.2: Got 503 from https://registry.yarnpkg.com/is-regex/1.1.2",
      ],
    ]);
  });

  it("sends Bearer token when requested", async () => {
    const token = "8dd79e01-d0a4-446b-b261-d6d6408227d2";
    const fetch: Fetcher = async (url, init) => {
      const { headers = new Headers() } = init ?? {};
      if (/is-regex\/1\.1\.2/.test(url)) {
        if (headers.get("Authorization") === `Bearer ${token}`) {
          return nodeFetch(url);
        } else {
          return new Response("", {
            status: headers.get("Authorization") ? 403 : 401,
            url,
          });
        }
      } else {
        return nodeFetch(url);
      }
    };
    const interaction = {
      warn: jest.fn<string, []>(),
    };
    const { completeRecording, assertScopesFinished } = await record(
      "yarn-lock1"
    );

    const lockV2 = fs.readFileSync(
      path.resolve(__dirname, "./__fixtures__/yarn-lock1.txt"),
      "utf-8"
    );
    const lockV1 = await convertLockfile(lockV2, {
      fetch,
      interaction,
      yarnrc: { npmAuthToken: token },
    });
    expect(lockV1).toMatchSnapshot();

    completeRecording();
    assertScopesFinished();

    expect(interaction.warn.mock.calls).toEqual([]);
  });
});
