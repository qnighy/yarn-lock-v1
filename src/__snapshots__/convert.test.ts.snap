// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertLockfile convers normal dependencies 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


call-bind@^1.0.2:
  version \\"1.0.2\\"
  resolved \\"https://registry.yarnpkg.com/call-bind/-/call-bind-1.0.2.tgz#b1d4e89e688119c3c9a903ad30abb2f6a919be3c\\"
  integrity sha512-7O+FbCihrB5WGbFYesctwmTKae6rOiIzmz1icreWJ+0aA7LJfuqhEso2T9ncpcFtzMQtzXf2QGGueWJGTYsqrA==
  dependencies:
    function-bind \\"^1.1.1\\"
    get-intrinsic \\"^1.0.2\\"

has-symbols@^1.0.1:
  version \\"1.0.2\\"
  resolved \\"https://registry.yarnpkg.com/has-symbols/-/has-symbols-1.0.2.tgz#165d3070c00309752a1236a479331e3ac56f1423\\"
  integrity sha512-chXa79rL/UC2KlX17jo3vRGz0azaWEx5tGqZg5pO3NUyEJVB17dMruQlzCCOfUvElghKcm5194+BCRvi2Rv/Gw==

is-regex@^1.1.2:
  version \\"1.1.2\\"
  resolved \\"https://registry.yarnpkg.com/is-regex/-/is-regex-1.1.2.tgz#81c8ebde4db142f2cf1c53fc86d6a45788266251\\"
  integrity sha512-axvdhb5pdhEVThqJzYXwMlVuZwC+FF2DpcOhTS+y/8jVq4trxyPgfcwIxIKiyeuLlSQYKkmUaPQJ8ZE4yNKXDg==
  dependencies:
    call-bind \\"^1.0.2\\"
    has-symbols \\"^1.0.1\\"
"
`;

exports[`convertLockfile converts confluent dependencies 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


\\"@babel/code-frame@^7.0.0\\":
  version \\"7.12.13\\"
  resolved \\"https://registry.yarnpkg.com/@babel/code-frame/-/code-frame-7.12.13.tgz#dcfc826beef65e75c50e21d3837d7d95798dd658\\"
  integrity sha512-HV1Cm0Q3ZrpCR93tkWOYiuYIgLxZXZFVG2VgK+MBWjUqZTundupbfx2aXarXuw5Ko5aMcjtJgbSs4vUGBS5v6g==
  dependencies:
    \\"@babel/highlight\\" \\"^7.12.13\\"

\\"@babel/helper-validator-identifier@^7.14.0\\":
  version \\"7.14.0\\"
  resolved \\"https://registry.yarnpkg.com/@babel/helper-validator-identifier/-/helper-validator-identifier-7.14.0.tgz#d26cad8a47c65286b15df1547319a5d0bcf27288\\"
  integrity sha512-V3ts7zMSu5lfiwWDVWzRDGIN+lnCEUdaXgtVHJgLb1rGaA6jMrtB9EmE7L18foXJIE8Un/A/h6NJfGQp/e1J4A==

\\"@babel/highlight@^7.0.0\\", \\"@babel/highlight@^7.12.13\\":
  version \\"7.14.0\\"
  resolved \\"https://registry.yarnpkg.com/@babel/highlight/-/highlight-7.14.0.tgz#3197e375711ef6bf834e67d0daec88e4f46113cf\\"
  integrity sha512-YSCOwxvTYEIMSGaBQb5kDDsCopDdiUGsqpatp3fOlI4+2HQSkTmEVWnVuySdAC5EWCqSWWTv0ib63RjR7dTBdg==
  dependencies:
    \\"@babel/helper-validator-identifier\\" \\"^7.14.0\\"
    chalk \\"^2.0.0\\"
    js-tokens \\"^4.0.0\\"

ansi-styles@^3.2.1:
  version \\"3.2.1\\"
  resolved \\"https://registry.yarnpkg.com/ansi-styles/-/ansi-styles-3.2.1.tgz#41fbb20243e50b12be0f04b8dedbf07520ce841d\\"
  integrity sha512-VT0ZI6kZRdTh8YyJw3SMbYm/u+NqfsAxEpWO0Pf9sq8/e94WxxOpPKx9FR1FlyCtOVDNOQ+8ntlqFxiRc+r5qA==
  dependencies:
    color-convert \\"^1.9.0\\"

chalk@^2.0.0:
  version \\"2.4.2\\"
  resolved \\"https://registry.yarnpkg.com/chalk/-/chalk-2.4.2.tgz#cd42541677a54333cf541a49108c1432b44c9424\\"
  integrity sha512-Mti+f9lpJNcwF4tWV8/OrTTtF1gZi+f8FqlyAdouralcFWFQWF2+NgCHShjkCb+IFBLq9buZwE1xckQU4peSuQ==
  dependencies:
    ansi-styles \\"^3.2.1\\"
    escape-string-regexp \\"^1.0.5\\"
    supports-color \\"^5.3.0\\"

color-convert@^1.9.0:
  version \\"1.9.3\\"
  resolved \\"https://registry.yarnpkg.com/color-convert/-/color-convert-1.9.3.tgz#bb71850690e1f136567de629d2d5471deda4c1e8\\"
  integrity sha512-QfAUtd+vFdAtFQcC8CCyYt1fYWxSqAiK2cSD6zDB8N3cpsEBAvRxp9zOGg6G/SHHJYAT88/az/IuDGALsNVbGg==
  dependencies:
    color-name \\"1.1.3\\"

color-name@1.1.3:
  version \\"1.1.3\\"
  resolved \\"https://registry.yarnpkg.com/color-name/-/color-name-1.1.3.tgz#a7d0558bd89c42f795dd42328f740831ca53bc25\\"
  integrity sha1-p9BVi9icQveV3UIyj3QIMcpTvCU=

escape-string-regexp@^1.0.5:
  version \\"1.0.5\\"
  resolved \\"https://registry.yarnpkg.com/escape-string-regexp/-/escape-string-regexp-1.0.5.tgz#1b61c0562190a8dff6ae3bb2cf0200ca130b86d4\\"
  integrity sha1-G2HAViGQqN/2rjuyzwIAyhMLhtQ=

has-flag@^3.0.0:
  version \\"3.0.0\\"
  resolved \\"https://registry.yarnpkg.com/has-flag/-/has-flag-3.0.0.tgz#b5d454dc2199ae225699f3467e5a07f3b955bafd\\"
  integrity sha1-tdRU3CGZriJWmfNGfloH87lVuv0=

js-tokens@^4.0.0:
  version \\"4.0.0\\"
  resolved \\"https://registry.yarnpkg.com/js-tokens/-/js-tokens-4.0.0.tgz#19203fb59991df98e3a287050d4647cdeaf32499\\"
  integrity sha512-RdJUflcE3cUzKiMqQgsCu06FPu9UdIJO0beYbPhHN4k6apgJtifcoCtT9bcxOpYBtpD2kCM6Sbzg4CausW/PKQ==

supports-color@^5.3.0:
  version \\"5.5.0\\"
  resolved \\"https://registry.yarnpkg.com/supports-color/-/supports-color-5.5.0.tgz#e2e69a44ac8772f78a1ec0b35b689df6530efc8f\\"
  integrity sha512-QjVjwdXIt408MIiAqCX4oUKsgU2EqAGzs2Ppkm4aQYbjm+ZEWEcW4SfFNTr4uMNZma0ey4f5lgLrkB0aX0QMow==
  dependencies:
    has-flag \\"^3.0.0\\"
"
`;

exports[`convertLockfile falls back to offline inference if fetch failed 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


call-bind@^1.0.2:
  version \\"1.0.2\\"
  resolved \\"https://registry.yarnpkg.com/call-bind/-/call-bind-1.0.2.tgz#b1d4e89e688119c3c9a903ad30abb2f6a919be3c\\"
  integrity sha512-7O+FbCihrB5WGbFYesctwmTKae6rOiIzmz1icreWJ+0aA7LJfuqhEso2T9ncpcFtzMQtzXf2QGGueWJGTYsqrA==
  dependencies:
    function-bind \\"^1.1.1\\"
    get-intrinsic \\"^1.0.2\\"

has-symbols@^1.0.1:
  version \\"1.0.2\\"
  resolved \\"https://registry.yarnpkg.com/has-symbols/-/has-symbols-1.0.2.tgz#165d3070c00309752a1236a479331e3ac56f1423\\"
  integrity sha512-chXa79rL/UC2KlX17jo3vRGz0azaWEx5tGqZg5pO3NUyEJVB17dMruQlzCCOfUvElghKcm5194+BCRvi2Rv/Gw==

is-regex@^1.1.2:
  version \\"1.1.2\\"
  resolved \\"https://registry.yarnpkg.com/is-regex/-/is-regex-1.1.2.tgz\\"
  dependencies:
    call-bind \\"^1.0.2\\"
    has-symbols \\"^1.0.1\\"
"
`;

exports[`convertLockfile sends Bearer token when requested 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


call-bind@^1.0.2:
  version \\"1.0.2\\"
  resolved \\"https://registry.yarnpkg.com/call-bind/-/call-bind-1.0.2.tgz#b1d4e89e688119c3c9a903ad30abb2f6a919be3c\\"
  integrity sha512-7O+FbCihrB5WGbFYesctwmTKae6rOiIzmz1icreWJ+0aA7LJfuqhEso2T9ncpcFtzMQtzXf2QGGueWJGTYsqrA==
  dependencies:
    function-bind \\"^1.1.1\\"
    get-intrinsic \\"^1.0.2\\"

has-symbols@^1.0.1:
  version \\"1.0.2\\"
  resolved \\"https://registry.yarnpkg.com/has-symbols/-/has-symbols-1.0.2.tgz#165d3070c00309752a1236a479331e3ac56f1423\\"
  integrity sha512-chXa79rL/UC2KlX17jo3vRGz0azaWEx5tGqZg5pO3NUyEJVB17dMruQlzCCOfUvElghKcm5194+BCRvi2Rv/Gw==

is-regex@^1.1.2:
  version \\"1.1.2\\"
  resolved \\"https://registry.yarnpkg.com/is-regex/-/is-regex-1.1.2.tgz#81c8ebde4db142f2cf1c53fc86d6a45788266251\\"
  integrity sha512-axvdhb5pdhEVThqJzYXwMlVuZwC+FF2DpcOhTS+y/8jVq4trxyPgfcwIxIKiyeuLlSQYKkmUaPQJ8ZE4yNKXDg==
  dependencies:
    call-bind \\"^1.0.2\\"
    has-symbols \\"^1.0.1\\"
"
`;
